### ISSUE TITLE
Auto Screen Rotation on Wayland

### ISSUE THREAD

**Issue by reulew** [linux-surface/linux-surface#903]
Physically rotating the tablet does not rotate the screen automatically. Some people said it was working for them without messing with anything but they were all using x11. Is there a way to fix auto screen rotation on Wayland?

### Environment

- Hardware model:  Surface Pro 7
- Kernel version:  5.18.16-arch1-1-surface
- Distribution:  Arch Linux (KDE Wayland)

**Comment by StollD** [comment]
Is `iio-sensor-proxy` installed?

**Comment by m1nicrusher** [comment]
In my case (SLS and SP8), I'll need to remove/collapse keyboard, namely go into tablet mode, to make "Screen Rotation Lock" option available in power menu. Maybe give it a try?

**Comment by reulew** [comment]
> Is `iio-sensor-proxy` installed?

Yes. I have tried both that and the git version.

> In my case (SLS and SP8), I'll need to remove/collapse keyboard, namely go into tablet mode, to make "Screen Rotation Lock" option available in power menu. Maybe give it a try?

I have tried removing the keyboard, going into tablet mode, and unchecking automatic orientation "Only In Tablet Mode".

**Comment by m1nicrusher** [comment]
> I have tried removing the keyboard, going into tablet mode, and unchecking automatic orientation "Only In Tablet Mode".

So does auto rotation work without keyboard? If so, I think it's normal. On my SP8, with keyboard attached, it will lock auto rotation automatically, which is a good thing. When you have your keyboard on you always want it to be towards you isn't it?

What desktop environment you're using? KDE?

**Comment by reulew** [comment]
> So does auto rotation work without keyboard?

No. I have tried keyboard and no keyboard with no success. I don't have a Microsoft type cover, just a regular keyboard.

> What desktop environment you're using? KDE?

Yes I'm using KDE Wayland (Plasma 5.25.4)

**Comment by NP-chaonay** [comment]
> Microsoft type cover

idk if tablet mode is triggered by plugging in and out of Microsoft type cover.

maybe waiting for who know about this telling you, I seem to have read some thing in (this repo)'s issues, it tell that Microsoft type cover may be required to triggering tablet mode either off or on, regular kbd dont do, but I not sure about it

**Comment by m1nicrusher** [comment]
> idk if tablet mode is triggered by plugging in and out of Microsoft type cover.

On my SP8, tablet mode is triggered by either attaching/removing Type Cover or deploying/folding TypeCover.

So I guess maybe in his case the tablet mode is kept off because he never have a TypeCover.

**Comment by reulew** [comment]
> So I guess maybe in his case the tablet mode is kept off because he never have a TypeCover.

I disabled this in settings so it shouldn't even matter. It should work both in tablet mode and not in tablet mode.

**Comment by NP-chaonay** [comment]
> > So I guess maybe in his case the tablet mode is kept off because he never have a TypeCover.
> 
> I disabled this in settings so it shouldn't even matter. It should work both in tablet mode and not in tablet mode.

oh I forget that you using KDE, so idk what is happening, yeah it should not matter

but are you sure that KDE is putting thing on tablet mode, and does auto rotation option appeared in KDE, I think it should have option to lock/unock screen rotate

**Comment by reulew** [comment]
> I think it should have option to lock/unock screen rotate

No lock screen rotate option appears for me. It doesnt make sense that you neet the typecover to get the tablet to rotate.

**Comment by NP-chaonay** [comment]
> > I think it should have option to lock/unock screen rotate
> 
> No lock screen rotate option appears for me. It doesnt make sense that you neet the typecover to get the tablet to rotate.

yes since you not using gnome wayland, however I have not using kde wayland/x11 on surface. So I could not solve your problem with much details.

On GNOME Shell Wayland. You have to put tablet mode to on before auto rotate/ osk  on touch work.
The way to make tablet mode is to triggered by typecover. so no typecover, you have to virtually implement tablet mode, if you DE dont have it I would safely assume that tablet mode is off. And if tabelt mode off can afffect the functionality to be turned off in GNOME Wayland, but the problem is Idk if that also disable what thing in KDE, so if you check that you will sure if virtual/real tablet mode is related to your problem.


If yes, then you have to implement one (through kernel module driver, I have make it 1) and set it enable. And see if behvaiour is normal or going enable for some functions.

If not, then I have nothing to say at now, since I dont use KDE on surface at this moment

**Comment by reulew** [comment]
> If yes, then you have to implement one (through kernel module driver, I have make it 1) and set it enable. And see if behvaiour is normal or going enable for some functions.

Are you saying the only way to enable tablet mode is to make my own custom driver?

**Comment by NP-chaonay** [comment]
> > If yes, then you have to implement one (through kernel module driver, I have make it 1) and set it enable. And see if behvaiour is normal or going enable for some functions.
> 
> Are you saying the only way to enable tablet mode is to make my own custom driver?

sorry for diffcult-writing of me I have readjusted; pls try to reread.  and answer my question (just answer for yours no need to put answer here)

if your answer is yes:  them yes, you have to make or get kernel driver for virtual set tablet mode.

**Comment by qzed** [comment]
> Are you saying the only way to enable tablet mode is to make my own custom driver?

Normally, that should not be necessary. It may be one workaround, but this seems to be an issue with the desktop environment not picking up the device as tablet correctly. The workaround with custom driver "fakes" a tablet mode switch so that the detection heuristics properly recognize it as a tablet/convertible. So I think the proper fix is to address this in the desktop env, i.e. KDE or at least check what's going wrong there.

**Comment by reulew** [comment]
> I'm using SP8, Is there a way I can disable the screen rotation?
> 
> Hardware: SP8 Kernel version: 6.0.6-surface Distribution: Ubuntu 22.04.1 LTS

@zed-wong Did any of the solutions listed [here](https://askubuntu.com/questions/1035209/how-to-turn-off-screen-rotation-in-ubuntu-18-04-lts) fix your problem? This issue is actually about the screen not auto-rotating when it should, not the other way around. 

As I understand, the original issue has yet to be rectified. The suggestion to make my own driver is slightly outrageous considering the normality of the problem. The kernel should support tablets with and without type covers natively.

**Comment by NP-chaonay** [comment]
@reulew 

> As I understand, the original issue has yet to be rectified.

what that bug could you post the linked to issue? 

> The kernel should support tablets with and without type covers natively.

yes (however idk the problem desc)

> about the screen

what you mean screen. The intel driver and Wayland/X11 seem have nothing to do with auto-rotation. I'm quite sure it is about screen-rotation and perhaps about device type detection algorithm/implmentation.

@zed-wong 

despite any bug, you can just power power menu top right of GNOME top bar and click lock screen rotation.

btw what you are doing is seem to be disable sensor across all that-API-compatible program/service.

@reulew  @zed-wong 
ok the known bug I known is:  `when tablet is enter (if I not wrong also appiled to when exiting) to tablet mode; the screen rotate activate once time (not all the time but just trigger auto-rotate 1x time)`

I'm sure it is GNOME bug since the iio-sensor-proxy.service is just provide sensor data through some API.

if the bug is on GNOME-shell, who control device rotation.

**Comment by NP-chaonay** [comment]
> @NP-chaonay I'm using Ubuntu 22 and it's running Wayland, the option you mentioned seemed to never appear unless the rotation was triggered. And even if the rotation was triggered and the option shows up, it won't work. It's really annoying so I just disabled the sensor.

oh that seem to be weird thing, ok this is my behaviour on SLS

```
[out of tablet mode]
- have to set screen orientation (ScrOrtt) manually, sensors not used for determine ScrOrtt
- the screen rotation unlock/lock menu disappeared from power menu
[trigger tablet mode or trigger that tablet mode is deactivated]
- sensors used for determine ScrOrtt rotation but just 1x time
[in tablet mode, auto-rotation enabled]
- cannot screen orientation (ScrOrtt) manually, sensors used for determine ScrOrtt
- the screen rotation lock menu appeared from power menu
[in tablet mode, auto-rotation enabled]
- can set screen orientation (ScrOrtt) manually, sensors not used for determine ScrOrtt
- - the screen rotation unlock menu appeared from power menu
```

have you upgraded the system and use linux-surface

@zed-wong

**Comment by NP-chaonay** [comment]
@reulew  ok I almost forgot your problem

If I have misunderstanding pls tell me

btw it seem that your problem and zed-wong is different 

yours is to make it auto-rotate
but zed-wrong is not disable it

**Comment by NP-chaonay** [comment]
> EDIT: In KDE, you might also need to detact the keyboard for rotation to work. ([StollD](https://github.com/StollD) msg)

oh is not that weird, should someone send feedback to KDE?

**Comment by NP-chaonay** [comment]
@reulew ok let me fix what I have confused

tablet mode driver or stuff now is does not matter

It is just synmtomp that KDE does not rotate according to sensor OR KDE does not found the sensors.

this is idk what to do about this. However...

Could you try testing this on GNOME-shell X11 (cannot using Wayland since it required tablet mode to be enabled in order to allow screen-rotation), if that works.

**Comment by cajhne** [comment]
After a ridiculous amount of googling, and turning up nothing but over-complicated and tedious hacks, I suddenly thought: "maybe there's an extension that does something like this".

Well, there is! It works perfectly to enable auto-rotation for Wayland in GNOME on my Lenovo Yoga:

https://extensions.gnome.org/extension/5389/screen-rotate/

Hope it helps you too!

**Comment by mikedebian** [comment]
Environment

    Hardware model: Surface Pro 7
    Kernel version: 6.4.7-arch1-1-surface
    Distribution: Arch Linux (KDE Wayland)

**Comment by Cal-Nine** [comment]
In KDE Plasma, my Auto-Rotate works with tablet mode enabled, but the auto-rotate controls that should be in Display & Monitor settings are hidden.

https://bugs.kde.org/show_bug.cgi?id=488139

"Touch Mode" can enable automatically so KDE can definitely recognise the tablet, and my auto-rotate works with Touch Mode so the accelerometer is working, I just can't access those hidden controls to turn on auto-rotate without also enabling Touch Mode.
